File: Algo/Medium/0063_Unique_Paths_2/Solution.py
from typing import *

## 2022-05-14
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[-1][-1] == 1:
            return 0
        
        num_paths = [[0 for _ in range(len(obstacleGrid[0]))] for i in range(len(obstacleGrid))]
        num_paths[0][0] = 1
        
        for i in range(len(num_paths)):
            for j in range(len(num_paths[i])):
                if obstacleGrid[i][j] == 1 or num_paths[i] == 0:
                    continue
                
                if i+1 < len(num_paths):
                    num_paths[i+1][j] += num_paths[i][j]
                if j+1 < len(num_paths[i]):
                    num_paths[i][j+1] += num_paths[i][j]
        
        return num_paths[-1][-1]
    

File: Algo/Medium/0063_Unique_Paths_2/Solution.cpp
// 2022-05-19
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if (obstacleGrid.front().front() == 1 || obstacleGrid.back().back() == 1)
            return 0;
        
        vector<vector<int>> reachable;
        for (int i=0; i<obstacleGrid.size(); i++) {
            vector<int> row;
            for (int j=0; j<obstacleGrid[i].size(); j++) {
                row.push_back(0);
            }
            reachable.push_back(row);
        }
        
        reachable[0][0] = 1;
        for (int i=0; i<obstacleGrid.size(); i++) {
            for (int j=0; j<obstacleGrid[i].size(); j++) {
                if (obstacleGrid[i][j] == 0 && reachable[0][0] > 0) {
                    if (i+1 < obstacleGrid.size() && obstacleGrid[i+1][j] == 0) {
                        reachable[i+1][j] += reachable[i][j];
                    }
                    if (j+1 < obstacleGrid[i].size() && obstacleGrid[i][j+1] == 0) {
                        reachable[i][j+1] += reachable[i][j];
                    }
                }
            }
        }
        return reachable.back().back();
    }
};